[BasicAA] Use nuw attribute of GEPs

hazzlim opened the pull request:
Use the nuw attribute of GEPs to prove that pointers do not alias, in cases matching the following:

```
   +                +                     +
   | BaseOffset     |   +<nuw> Indices    |
   ---------------->|-------------------->|
   |-->VLeftSize    |                     |------->VRightSize
  LHS                                    RHS
```

If the difference between pointers is Offset +<nuw> Indices (the variable indices all come from nuw GEPs) then we know that the addition does not wrap the pointer index type (add nuw) and the constant Offset is a lower bound on the distance between the pointers. We can then prove NoAlias via Offset u>= VLeftSize.

hazzlim committed:
[BasicAA] Use nuw attribute of GEPs

Use the nuw attribute of GEPs to prove that pointers do not alias, in
cases matching the following:

   +                +                     +
   | BaseOffset     |   +<nuw> Indices    |
   ---------------->|-------------------->|
   |-->VLeftSize    |                     |------->VRightSize
  LHS                                    RHS

If the difference between pointers is Offset +<nuw> Indices (the variable
indices all come from nuw GEPs) then we know that the addition does not
wrap the pointer index type (add nuw) and the constant Offset is a lower
bound on the distance between the pointers. We can then prove NoAlias via
Offset u>= VLeftSize.

david-arm commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -1233,6 +1232,32 @@ AliasResult BasicAAResult::aliasGEP(
+
+    bool IndicesFromRight = DecompGEP1.VarIndices.front().IsNegated == Swapped;
+    if (IndicesFromRight && DecompRight.NWFlags->hasNoUnsignedWrap())
+      if (!VLeftSize.isScalable() && VLeftSize.hasValue() &&
```
nit: I must admit I haven't looked into the LocationSize class in great detail, but  it looks like asking the question `isScalable` only makes sense if `hasValue` returns true? So even though there is nothing technically wrong with the code it might be more logical to reorder the checks, i.e.

```
if (VLeftSize.hasValue() && !VLeftSize.isScalable()
```

However, I could have misunderstood the semantics!

hazzlim replied:
I think you are right - this is the order of the check above for example - I've changed to this order here.

As a related note - I think that this could be extended to support cases where `VLeftSize` is scalable, in the same way as this case for constant-only offsets: https://github.com/llvm/llvm-project/blob/9f4a25e2a7cd176bd4f946dc651bc18c7a2e8c92/llvm/lib/Analysis/BasicAliasAnalysis.cpp#L1195

However it seemed to make sense to leave this to a later commit!

hazzlim committed:
Reorder checks (Addressing review comment)

hazzlim commented:
Gentle ping @nikic if you have any comments on this?

nikic reviewed (commented):
I don't think the approach here is quite right. Consider this case: https://alive2.llvm.org/ce/z/TvPTRV

Your current implementation will say that the pointers don't alias, but they may. We need to be careful about how the GEP subtraction affects whether NUW still holds afterwards.

nikic commented:
By the way, I put up https://github.com/llvm/llvm-project/pull/102141 as a complement to this, which handles the simpler case where you have both nusw and nuw (or inbounds and nuw). I think many use cases will be covered by that one, but handling for the "nuw only" case is still desirable, it's just trickier.

nikic commented:
> Thank you for reviewing and pointing out the issues with this patch - I see that I was not sufficiently considering how the GEP subtraction works in cases such as that example. I guess for this approach to work we'd need to prove that the constant offset between the GEPs does not cross the address space boundary, and its not obvious to me how to do this (without inbounds/nusw).

I think the necessary condition for the constant offset is just that the LHS offset is larger than the RHS offset, so that the subtraction does not overflow (in an unsigned sense).

An approach to make this work could be:

 * Swap LHS/RHS so that the larger GEP (in the number of variables) is on the left. I think doing this ahead of time is safer than trying to reason about IsNegated, because the IsNegated flag may get lost.
 * Subtract LHS - RHS as usual, but need to clear the NUW flag in some cases now.
 * If LHS.Offset ult RHS.Offset, clear NUW.
 * For variable indices, if LHS.Scale ult RHS.Scale, clear NUW.

hazzlim committed:
Drop nuw flag in some cases when subtracting GEPs

hazzlim commented:
Hi @nikic 

Thank you for pointing me in the right direction to make this work. I have pushed an updated implementation that drops the NUW flag in the cases you outlined.

>Swap LHS/RHS so that the larger GEP (in the number of variables) is on the left. I think doing this ahead of time is safer than trying to reason about IsNegated, because the IsNegated flag may get lost.

I had a look at doing this, but it seems like it would require changing a lot of code in `aliasGEP` to make this change work. Hopefully the approach I've taken here seems reasonable - as far as I understand it we should be okay to reason about `IsNegated || Scale.isNegative()` as the IsNegated flag is only lost when we are subtracting the Scale. However if you'd still prefer to Swap LHS/RHS I'd be happy to make that change.

hazzlim committed:
Swap order of GEPs to simplify things

hazzlim commented:
Sorry, you can ignore the previous update - I was missing how straightforward it would be to swap the GEPs ahead of time.

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -552,17 +552,17 @@ struct BasicAAResult::DecomposedGEP {
+  // Nowrap flags common to all GEP operations involved in expression.
   // (std::nullopt iff expression doesn't involve any geps)
-  std::optional<bool> InBounds;
+  std::optional<GEPNoWrapFlags> NWFlags;
```
```suggestion
  GEPNoWrapFlags NWFlags = GEPNoWrapFlags::all();
```
Should make things a bit clearer. I don't think the optional adds anything here...

hazzlim replied:
Good point - I've made this initialize with `GEPNoWrapFlags::none()`, with some logic in `DecomposeGEPExpression` so that we initialize on the first GEP we encounter - as this seemed to make sense to avoid ever having DecompGEP2 retain `GEPNoWrapFlags::all()` in the case where the decomposition of `V2` does not encounter any GEPs. 

This doesn't seem to be an issue at present, but seems cleaner to make sure the flags are always correct in this way?

nikic replied:
No GEP would be equivalent to a zero-offset GEP, for which all of inbounds, nusw, and nuw always hold. In the past, it was unclear whether inbounds held for zero-offset GEPs, but this has been decided now. As such, initializing this to all() is correct, and should make things a bit simpler because we don't need the separate flag.

hazzlim replied:
Ah yes, I can see this case is actually spelled out in LangRef, thanks for pointing that out. Updated to initialize with all() and remove the now unnecessary conditional.

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -552,17 +552,17 @@ struct BasicAAResult::DecomposedGEP {
   void print(raw_ostream &OS) const {
-    OS << "(DecomposedGEP Base=" << Base->getName()
-       << ", Offset=" << Offset
+    OS << "(DecomposedGEP Base=" << Base->getName() << ", Offset=" << Offset
```
I'd put the `nuw` before `Offset`. A bit odd to have it between offset and indices.

hazzlim replied:
Done, and added `inbounds` here also for symmetry.

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -1159,8 +1163,6 @@ AliasResult BasicAAResult::aliasGEP(
 
     // Initialize for Off >= 0 (V2 <= GEP1) case.
-    const Value *LeftPtr = V2;
-    const Value *RightPtr = GEP1;
```
Not related to this change -- I've landed this separately in https://github.com/llvm/llvm-project/commit/228753290b7426bf3eda7c4397aa74f0c67d07d7.

hazzlim replied:
Nice, thank you - rebased on this.

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -1239,6 +1240,20 @@ AliasResult BasicAAResult::aliasGEP(
+  //    +                +                     +
+  //    | BaseOffset     |   +<nuw> Indices    |
+  //    ---------------->|-------------------->|
+  //    |-->VLeftSize    |                     |-------> VRightSize
```
```suggestion
  //    |-->V2Size    |                     |-------> V1Size
```
I think? WOuld be good to have the names here line up with the variable names.

hazzlim replied:
Done.

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -1239,6 +1240,20 @@ AliasResult BasicAAResult::aliasGEP(
+  //   LHS                                    RHS
+  if (!DecompGEP1.VarIndices.empty() &&
+      DecompGEP1.NWFlags->hasNoUnsignedWrap() && V2Size.hasValue() &&
+      !V2Size.isScalable() && DecompGEP1.Offset.sge(V2Size.getValue()))
```
```suggestion
      !V2Size.isScalable() && DecompGEP1.Offset.uge(V2Size.getValue()))
```
As we're working with nuw?

hazzlim replied:
Good point - done.

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -1843,6 +1858,11 @@ bool BasicAAResult::isValueEqualInPotentialCycles(const Value *V,
+  // Drop nuw flag from GEP if subtraction of constant offsets overflows in an
+  // unsigned sense.
+  if (DestGEP.Offset.ult(SrcGEP.Offset))
+    DestGEP.NWFlags = DestGEP.NWFlags->withoutNoUnsignedWrap();
```
(Another reason to drop the std::optional: It's not really clear that this code here can't access nullopt).

nikic reviewed (commented):
Thanks, this looks much better!

hazzlim committed:
Address review comments

- Remove the std::optional for GEPNoWrapFlags
- Reorder `nuw` in DecomposedGEP.print(), and add inbounds
- Refactor comment to use same names as Size variable names
- Use uge for comparison

hazzlim committed:
Initialize DecomposedGEP.NWFlags to GEPNoWrapFlags::all()

nikic reviewed (approved):
LGTM

hazzlim committed:
Address final review comments

- Reorder DecomposedGEP print to match getelementptr syntax
- Remove unnecessary datalayout string from test
